docs: Add VM support GSG (#13940)
cilium-team/cilium: Documentation/gettingstarted/external-workloads.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _external_workloads:

.. _gs_external_workloads:

*************************************************
Setting up Support for External Workloads (beta)
*************************************************

This is a step-by-step guide on how to add external workloads (such as
VMs) in to your Kubernetes cluster and to enforce security policies to
restrict access.

.. include:: ../beta.rst

Prerequisites
#############

* Cilium must be configured to Kubernetes for identity allocation

>   (``identityAllocationMode`` set to ``crd``).

Author: Mention that this is the default mode for new installations.

enable a cluster job to create TLS secrets instead. This way the CA
cert secrets are reused as long as they are not removed from your
cluster:

.. parsed-literal ::

    helm install cilium |CHART_RELEASE|                     \\
      --namespace kube-system                               \\
      --set externalWorkloads.enabled=true                  \\
      --set clustermesh.apiserver.tls.auto.method=cronJob

Secondly, you can use your own CA certs. You can either create the
``clustermesh-apiserver-ca-cert`` secret in your Cilium install
namespace (e.g., ``kube-system``) yourself, or pass the CA cert and
key to helm:

.. parsed-literal ::

    helm install cilium |CHART_RELEASE|                     \\
      --namespace kube-system                               \\
      --set externalWorkloads.enabled=true                  \\
      --set clustermesh.apiserver.tls.auto.method=cronJob   \\
      --set clustermesh.apiserver.tls.ca.cert="<< base64 encoded CA certificate >>"   \\
      --set clustermesh.apiserver.tls.ca.key="<< base64 encoded CA key >>"

You can also allow the cronJob create the secrets and then save the
generated CA certs to be used later:

.. code:: bash

>     kubectl -n kube-system get secret clustermesh-apiserver-ca-cert -o yaml > CA.secret

Author: `> clustermesh-apiserver-ca-cert.yaml` instead

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _external_workloads:

.. _gs_external_workloads:

*************************************************
Setting up Support for External Workloads (beta)
*************************************************

This is a step-by-step guide on how to add external workloads (such as
VMs) in to your Kubernetes cluster and to enforce security policies to
restrict access.

.. include:: ../beta.rst

Prerequisites
#############

> * Cilium must be configured to Kubernetes for identity allocation

Reviewer: ```suggestion
* Cilium must be configured to use Kubernetes for identity allocation
```

  (``identityAllocationMode`` set to ``crd``). This is the default
  for new installations.

* External workloads need to run recent enough kernel for k8s service
  access from the external host to work, see :ref:`host-services` for

>   details.

Reviewer: ```suggestion
* External workloads must run a recent enough kernel for k8s service
  access from the external host to work, see :ref:`host-services` for
  details.
```

* External workloads must have IP connectivity with the nodes in your
  cluster. This requirement is typically met by establishing peering
  or VPN tunnels between the networks of the nodes of your cluster and
  your external workloads.

* All external workloads must have a unique IP address assigned
  them. Node IPs of such nodes and your clusters may not conflict with

>   each other.

Reviewer: ```suggestion
* All external workloads must have a unique IP address assigned
  them. Node IPs of such nodes and your clusters must not conflict with
  each other.
```

externalWorkloads.enabled=true`` to ``helm install`` when installing
or updating Cilium:

.. parsed-literal ::

    helm install cilium |CHART_RELEASE|         \\
      --namespace kube-system                   \\
      --set externalWorkloads.enabled=true

This will add a deployment for ``clustermesh-apiserver`` into your
cluster, as well as the related cluster resources, such as TLS
secrets.

.. _gs_external_workloads_tls_configuration:

TLS configuration
=================

By default TLS secrets are created by helm. This has the downside that
each time you run ``helm`` the TLS secrets will be re-created, forcing
each external workload to be reconfigured to be able to connect to
your cluster. There are two ways to get around this. Firstly, you can
enable a cluster job to create TLS secrets instead. This way the CA
cert secrets are reused as long as they are not removed from your
cluster:

.. parsed-literal ::

    helm install cilium |CHART_RELEASE|                     \\
      --namespace kube-system                               \\
      --set externalWorkloads.enabled=true                  \\
      --set clustermesh.apiserver.tls.auto.method=cronJob

> Secondly, you can use your own CA certs. You can either create the

Reviewer: ```suggestion
Alternatively, you can use your own CA certs. You can either create the
```

Reviewer: "Firstly ... Secondly ..." sounds like an order of operations, not alternatives.

``clustermesh-apiserver-ca-cert`` secret in your Cilium install
namespace (e.g., ``kube-system``) yourself, or pass the CA cert and
key to helm:

.. parsed-literal ::

    helm install cilium |CHART_RELEASE|                     \\
      --namespace kube-system                               \\
      --set externalWorkloads.enabled=true                  \\
      --set clustermesh.apiserver.tls.auto.method=cronJob   \\
      --set clustermesh.apiserver.tls.ca.cert="<< base64 encoded CA certificate >>"   \\
      --set clustermesh.apiserver.tls.ca.key="<< base64 encoded CA key >>"

You can also allow the cronJob create the secrets and then save the
generated CA certs to be used later:

.. code:: bash

>     kubectl -n kube-system get secret clustermesh-apiserver-ca-cert -o yaml > clustermesh-apiserver-ca-cert.yaml

Reviewer: Is this a tip for the first or second approach? Without following the guide myself, it's not entirely obvious why I might want to do this or how I will use it later on.

Author: Reworded as: _"You can also allow the cronJob create the secrets using the first approach above and then save the generated CA certs to be used with the second approach above:"_

   
.. _gs_external_workloads_expose_clustermesh_apiserver:

Expose cluster to external workloads
====================================

``clustermesh-apiserver`` must be exposed to the external
workloads. By default this is done using a HostPort service on

> port 32379. It is also possible to use the ``LoadBalancer`` service

Reviewer: Is it a HostPort or a NodePort? The first tab below says NodePort.

Author: NodePort, fixed.

      --set clustermesh.apiserver.service.loadBalancerIP="xxx.xxx.xxx.xxx"

    ``xxx.xxx.xxx.xxx`` must be an IP from ``gke-vip-subnet``.

    If not using a pre-set IP, get the service IP with:

    .. code:: bash

      kubectl -n kube-system get svc clustermesh-apiserver -o jsonpath='{.status.loadBalancer.ingress[0].ip}{"\n"}'

  .. group-tab:: AWS

    .. parsed-literal::

      --set clustermesh.apiserver.service.type=LoadBalancer
      --set clustermesh.apiserver.service.annotations."service\.beta\.kubernetes\.io/aws-load-balancer-internal"="true"

    If not using a pre-set IP, get the service IP with:

    .. code:: bash

      kubectl -n kube-system get svc clustermesh-apiserver -o jsonpath='{.status.loadBalancer.ingress[0].ip}{"\n"}'

.. note::

   Make sure that you use the namespace in which cilium is
   running. Depending on which installation method you chose, this
   could be ``kube-system`` or ``cilium``.

Tell your cluster about external workloads
==========================================

To allow an external workload to join your cluster, the cluster must
be informed about each such workload. This is done by adding a

> ``CiliumExternalWorkload`` (CEW) CRD for each external workload. CEW

Reviewer: ```suggestion
``CiliumExternalWorkload`` (CEW) resource for each external workload. The CEW
```

CRD specifies the name, namespace, and labels for the workload. For
now you must also allocate a small IP CIDR that must be unique to the
workload. For example, for a VM named ``runtime`` that is to join the
``default`` namespace, you could create a file ``runtime.yaml`` with
the following contents:

> .. parsed-literal ::

Reviewer: ```suggestion
.. code-block:: yaml
```

    apiVersion: cilium.io/v2
    kind: CiliumExternalWorkload
    metadata:
      name: runtime
      namespace: default
      labels:
        app: runtime
    spec:

>       ipv4-alloc-cidr: 10.192.1.0/30

Reviewer: It'd be nice to include something like this under `examples/`, use an `include` statement to include it, and ideally use the same example in the CI so we can validate that the example works.

(Reasonable as a followup)

Apply this with:

.. code:: bash

    kubectl apply -f runtime.yaml

Extract the TLS keys for external workloads
===========================================

Cilium control plane performs TLS based authentication and encryption.
For this purpose, the TLS keys and certificates of
``clustermesh-apiserver`` need to be made available to external
workload that wish to join the cluster.

Extract the TLS certs from your k8s cluster:

.. code:: bash

    ../../contrib/k8s/extract-external-workload-certs.sh

This saves the certs (``ca.crt``, ``tls.crt``, ``tls.key``) to the
current directory. These files need to be copied to your external
workload.

Install and configure Cilium on external workloads
##################################################

Log in to the external workload. First make sure the hostname matches
the name used in the CiliumExternalWorkload CRD:

.. code:: bash

    hostname

By now you should be able to find the corresponding CRD in your k8s
cluster (``<name>`` is the output from ``hostname`` above):

.. code:: bash

>     kubectl get cew <name>

Reviewer: It's not immediately clear, but these two commands would be run in different contexts right? Because the external workload probably doesn't have `kubectl` installed.

Next, copy the saved TLS certs from your kubectl command line and
``contrib/k8s/install-external-workload.sh`` from the Cilium release

> you downloaded in the beginning to the external workload. Then,

Reviewer: I didn't see an explicit command to download this file, and note that branch versions of the docs will use `helm install ...`, they will *not* download a Cilium tarball.

I suggest we add a `.. parsed-literal::` `curl` command to pull the install-external-workload.sh script from the current branch.

Author: Missed that difference, changed to download the script from the master branch.

assuming they are in the same directory:

.. tabs::
  .. group-tab:: NodePort

    .. parsed-literal ::

      CLUSTER_ADDR=<node-ip> CILIUM_IMAGE=cilium/cilium:|IMAGE_TAG| ./install-external-workload.sh

    ``<node-ip>`` is the node IP you extracted from the k8s cluster above.

  .. group-tab:: LoadBalancer

    .. parsed-literal ::

      CLUSTER_ADDR=<load-balancer-ip> CILIUM_IMAGE=cilium/cilium:|IMAGE_TAG| ./install-external-workload.sh

    ``<load-balancer-ip>`` is the load balancer IP you extracted from the k8s cluster above.

This command launches the Cilium agent in a docker container named
``cilium`` and copies the ``cilium`` CLI to your host. This needs
``sudo`` permissions, so you may be asked for a password.

This command waits until the node has been connected to the cluster
and the cluster services are available. Then it re-configures
``/etc/resolv.conf`` with the IP address of the ``kube-dns`` service.

.. note::

    If your node has multiple IP addresses you may need to tell Cilium

>     agent which to use. To this end add ``HOST_IP=<ip-address>`` to

Reviewer: ```suggestion
    agent which IP to use. To this end add ``HOST_IP=<ip-address>`` to
```

    the beginning of the command line above.

Verify basic connectivity
=========================

Next you can check the status of the Cilium agent in your external workload:

.. code:: bash

    cilium status

You should see something like:

.. parsed-literal::

    etcd: 1/1 connected, lease-ID=7c02748328e75f57, lock lease-ID=7c02748328e75f59, has-quorum=true: 192.168.36.11:32379 - 3.4.13 (Leader)
    ...

Check that cluster DNS works:

.. code:: bash

    nslookup clustermesh-apiserver.kube-system.svc.cluster.local

Inspecting status changes in the cluster
========================================

The following command in your cluster should show the external workload IPs and their Cilium security IDs:

.. code:: bash

    kubectl get cew

External workloads should also be visible as Cilium Endpoints:

.. code:: bash

    kubectl get cep

Apply Cilium Network Policy to enforce traffic from external workloads
======================================================================

> Ping the backend IP of ``clustermesh-apiserver`` service to verify connectivity:

Reviewer: ```suggestion
From the external workload, ping the backend IP of ``clustermesh-apiserver`` service to verify connectivity:
```

.. code:: bash

    ping $(cilium service list get -o jsonpath='{[?(@.spec.flags.name=="clustermesh-apiserver")].spec.backend-addresses[0].ip}')

The ping should keep running also when the following CCNP is applied in your cluster:

.. parsed-literal::

    apiVersion: cilium.io/v2
    kind: CiliumClusterwideNetworkPolicy
    metadata:
      name: test-ccnp
      namespace: kube-system
    spec:
      endpointSelector:
        matchLabels:
          k8s-app: clustermesh-apiserver
      ingress:
      - fromEndpoints:
        - matchLabels:
            io.kubernetes.pod.name: runtime
      - toPorts:
        - ports:
          - port: "2379"

>             protocol: TCP

Reviewer: Is this intended as two rules or just one?

Author: Two rules, latter needed for etcd itself.

The ping should stop if you delete these lines from the policy (e.g., ``kubectl edit ccnp test-ccnp``):

.. parsed-literal::

      - fromEndpoints:
        - matchLabels:

>             io.kubernetes.pod.name: runtime

Reviewer: While I think this does technically prove that security enforcement operates with the external workload being handled via a rich Identity, IMO it doesn't demonstrate the intended use cases for external workloads very well (eg, allowing access towards an external database, etc.) I wonder if we could come up with an example that better motivates the feature.

This guide also stops abruptly after this, it would be nice to have some summary text to establish what has now happened and what is possible.

Author: Added clean-up and conclustions sections.
